### PR TITLE
Add some framework for SSL tests and refactor makefile a bit

### PR DESCRIPTION
**Pull request by webknjaz**
* **What kind of change does this PR introduce?**
  - [x] bug fix
  - [ ] feature
  - [ ] docs update
  - [x] tests/coverage improvement
  - [x] refactoring
  - [ ] other

* **What is the related issue number (starting with `#`)**
#6 (`cheroot.ssl.pyopenssl` is now importable from under Python 3, but still doesn't work)
#95 (origins of testing SSL, added `trustme` for generation of CA and certs in runtime)

* **What is the current behavior?** (You can also link to an open issue here)
Mess in `cheroot.makefile`, `cheroot.ssl.pyopenssl` import fails under Python 3 (trying to inherit class from function :see_no_evil:), no SSL tests, no way to generate test certificates.

* **What is the new behavior (if this is a feature change)?**
New test for adapters in the wild, makefile has separately importable `StreamWriter` and `StreamReader`, 

* **Other information**:
https://gitter.im/cherrypy/cherrypy?at=5b871411c2bd5d117aefa6db

* **Checklist**:

  - [≈] I think the code is well written
  - [≈] I wrote [good commit messages][1]
  - [≈] I have [squashed related commits together][2] after the changes have been approved
  - [ ] Unit tests for the changes exist
  - [x] Integration tests for the changes exist (if applicable)
  - [x] I used the same coding conventions as the rest of the project
  - [x] The new code doesn't generate linter offenses
  - [ ] Documentation reflects the changes
  - [x] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences

[1]: http://chris.beams.io/posts/git-commit/
[2]: https://github.com/todotxt/todo.txt-android/wiki/Squash-All-Commits-Related-to-a-Single-Issue-into-a-Single-Commit

---
This change is [Reviewable](https://reviewable.io/reviews/cherrypy/cheroot/109)

### CODE REVIEW

**Comment by webknjaz** [comment]
> [WinError 10049] The requested address is not valid in its context

This error is happening under AppVeyor for some odd reason.

> address = ('0.0.0.0', 1191)

That's probably because connecting to `0.0.0.0` instead of localhost..

**Comment by webknjaz** [comment]
https://stackoverflow.com/a/28230165/595220

File: cheroot/makefile.py [hunk 1]
```diff
@@ -384,4 +376,27 @@ def readline(self, size=-1):
...
 
 
-MakeFile = MakeFile_PY2 if six.PY2 else MakeFile_PY3
+if six.PY3:
+    class StreamReader(io.BufferedReader):
+        """Socket stream reader."""
+
+        def __init__(self, sock, mode='r', bufsize=io.DEFAULT_BUFFER_SIZE):
+            """Initialize socket stream reader."""
+            super().__init__(socket.SocketIO(sock, mode), bufsize)
```
Comment (webknjaz) [inline cheroot/makefile.py]: _(This is always evaluated under Python 3, so no need to use Python 2 compatible `super()` syntax)_

**Comment by webknjaz** [comment]
Makefile related stuff: https://github.com/urllib3/urllib3/blob/master/src/urllib3/contrib/securetransport.py#L293

**Comment by webknjaz** [comment]
@jeffvanvoorst take into account that there's a lot of ___very old___ code, which we sometimes have trouble understanding completely. Among that, there's lots of backwards compatibility hacks. AFAIR `socket._fileobject` (missing in Python 3) was needed for legacy Python 2 versions (a year or two ago you could still find some Python 2.3 compat/backports stuff).
That's what makes it hard to read/support/change. And that's why we _want_ to refactor it and need help in figuring out how to do this harmlessly.

**Comment by webknjaz** [comment]
Alright, it looks like [`python-future`](https://pypi.org/project/future) has a backport for that: https://github.com/PythonCharmers/python-future/blob/master/src/future/backports/socket.py#L268-L384

@jaraco I think it'd be a good idea to adopt it. What do you think? I'd simplify our compatibility layers a lot!

**Comment by jaraco** [comment]
I've evaluated python-future in the past, and it seems like a pretty robust solution. The fact that it monkey patches the standard library has caused me problems in the past when I would be testing for Python 2/3 compatibility in an environment that had python-future, my results were influenced by the presence of that library, which left me with a slight distaste for it. But in general, I've seen good things come from it.

My main reluctance stems from our existing use of six. I imagine the two can interoperate, though it's not mentioned in the [narrative on the topic](http://python-future.org/faq.html#what-is-the-relationship-between-future-and-six).

However, given the potential benefits (more single-code base aligned to Python 3), I'm very much in favor, especially if this breathes some extra bit of life into the last Python 2 releases.

File: cheroot/makefile.py [hunk 2]
```diff
@@ -59,6 +51,16 @@ def __init__(self, *args, **kwargs):
...
+        self._refcount = 0
+
+    def _reuse(self):
+        self._refcount += 1
+
+    def _drop(self):
+        if self._refcount < 0:
+            self.close()
+        else:
+            self._refcount -= 1
```
Comment (jaraco) [inline cheroot/makefile.py]: This refcounting seems wrong to me. I would expect a refcount to be set to 1 when there's one object referencing it, and 0 when there are no objects referencing it. (not 0 when there's one, and -1 when there's none).

File: cheroot/ssl/builtin.py [hunk 3]
```diff
@@ -200,4 +199,5 @@ def env_dn_dict(self, env_prefix, cert_value):
 
     def makefile(self, sock, mode='r', bufsize=DEFAULT_BUFFER_SIZE):
         """Return socket file object."""
-        return MakeFile(sock, mode, bufsize)
+        cls = StreamReader if 'r' in mode else StreamWriter
```
Comment (jaraco) [inline cheroot/ssl/builtin.py]: It feels like this logic duplicates logic in makefile.MakeFile (Python 3 version). Ideally, this decision would be made in one place. Though, if one of these methods is deprecated, that's fine.

File: cheroot/ssl/pyopenssl.py [hunk 4]
```diff
@@ -102,17 +102,103 @@ def _safe_call(self, is_reader, call, *args, **kwargs):
...
+        )
+
+
+class SSLFileobjectStreamWriter(StreamWriter):
+    """SSL file object attached to a socket object."""
+
+    ssl_timeout = 3
+    ssl_retry = .01
+
+    def _safe_call(self, is_reader, call, *args, **kwargs):
```
Comment (jaraco) [inline cheroot/ssl/pyopenssl.py]: This function seems to add a lot of repetition. I suggest extracting the common aspects of _safe_call into a mix-in for each of these SSLFileobject classes.

**Comment by webknjaz** [comment]
@jaraco does it run monkey-patching by default? I thought it should be an explicit call for that.
Can we just import tiny things from there skipping the monkey-patching?

File: cheroot/ssl/pyopenssl.py (hunk 4 above)
Comment (webknjaz) [reply cheroot/ssl/pyopenssl.py]: Yeah, I just did this quick and dirty, I'm also in favor of having a mixin. Just didn't get to actually doing it.

File: cheroot/makefile.py (hunk 2 above)
Comment (webknjaz) [reply cheroot/makefile.py]: I stole that from https://github.com/urllib3/urllib3/blob/master/src/urllib3/contrib/securetransport.py#L670-L677
and it still doesn't seem to help. I'd postpone solving this for now.

File: cheroot/ssl/builtin.py (hunk 3 above)
Comment (webknjaz) [reply cheroot/ssl/builtin.py]: I decided to delay this refactoring. I also don't like the inability to remove `MakeFile` from `cheroot.server` module completely...

**Comment by jaraco** [comment]
> does [python-future] run monkey-patching by default?

Perhaps not. I guess what I observed was that `import builtins` only works on Python 2 when `python-future` is installed, and it was that behavior that confused me. Now that I look at it, I'm not seeing monkey-patching as I'd imagined:

```
~ $ python2.7 -m rwt future
Collecting future
  Using cached https://files.pythonhosted.org/packages/00/2b/8d082ddfed935f3608cc61140df6dcbf0edea1bc3ab52fb6c29ae3e81e85/future-0.16.0.tar.gz
Building wheels for collected packages: future
  Running setup.py bdist_wheel for future ... done
  Stored in directory: /Users/jaraco/Library/Caches/pip/wheels/bf/c9/a3/c538d90ef17cf7823fa51fc701a7a7a910a80f6a405bf15b1a
Successfully built future
Installing collected packages: future
Successfully installed future-0.16.0
Python 2.7.15 (default, Jun 20 2018, 13:24:23)
[GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import builtins
>>> builtins.str
<class 'future.types.newstr.newstr'>
>>> str
<type 'str'>
```

So in that case, one does have to explicitly `from builtins import str` to override the default `str` on Python 2 (same as we do with six).

So my fears are assuaged (and not justified), so let's try it.

**Comment by jaraco** [comment]
I should also say - until this example, I've found no instances where use of `python-future` would provide functionality that `six` does not, so I've had little motivation to try it out. Glad to see it adopted here.

**Comment by webknjaz** [comment]
Yeah, the most interesting thing to me is its collection of backports, unfortunately, this is not fully documented. I've found `socket.SocketIO` by doing googling and nothing mentioning it in docs. Then I went further, looked at the source and realized that there's quite a number of useful backports shipped with this dist.

**Comment by webknjaz** [comment]
This pull request **introduces 2 alerts** when merging c6b37adee512c91e0cfdce5eaea23a197e2575e3 into 5edf9cb1cd4ebc8775971909a037c4d9dcd6688b - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-1fb8351fa5fad7fa852f1ed5f2a7d66262c89602)

**new alerts:**

* 2 for Conflicting attributes in base classes

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 3 alerts** when merging a14410c38d96e9022c046417917efebc32f5c9ce into 5edf9cb1cd4ebc8775971909a037c4d9dcd6688b - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-4c7dab48e63a7aa66aa39f135dc2ef97d8c0ad21)

**new alerts:**

* 2 for Conflicting attributes in base classes
* 1 for Non\-callable called

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 3 alerts** when merging 69dfb979cf62660345530e3a504c7edf20b93959 into 5edf9cb1cd4ebc8775971909a037c4d9dcd6688b - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-78fea241a453b44ce4b15edbdc3c2264ef96c3d6)

**new alerts:**

* 2 for Conflicting attributes in base classes
* 1 for Non\-callable called

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: cheroot/server.py [hunk 5]
```diff
@@ -87,7 +87,7 @@
 from . import errors, __version__
 from ._compat import bton, ntou
 from .workers import threadpool
-from .makefile import MakeFile
+from .makefile import MakeFile, StreamWriter
```
Comment (webknjaz) [inline cheroot/server.py]: *(It annoys me that I've failed to remove `MakeFile` at all, but that's all I can do for now)*

File: cheroot/ssl/pyopenssl.py [hunk 6]
```diff
@@ -126,33 +188,6 @@ def __init__(self, *args):
...
 
-    for f in ('get_context', 'pending', 'send', 'write', 'recv', 'read',
-              'renegotiate', 'bind', 'listen', 'connect', 'accept',
-              'setblocking', 'fileno', 'close', 'get_cipher_list',
-              'getpeername', 'getsockname', 'getsockopt', 'setsockopt',
-              'makefile', 'get_app_data', 'set_app_data', 'state_string',
-              'sock_shutdown', 'get_peer_certificate', 'want_read',
-              'want_write', 'set_connect_state', 'set_accept_state',
-              'connect_ex', 'sendall', 'settimeout', 'gettimeout'):
-        exec("""def %s(self, *args):
```
Comment (webknjaz) [inline cheroot/ssl/pyopenssl.py]: I didn't like that it was `exec`ing things

File: cheroot/ssl/builtin.py (hunk 3 above)
Comment (webknjaz) [reply cheroot/ssl/builtin.py]: Oh, and I'm also trying hard to decouple those things.

Comment (webknjaz) [reply cheroot/ssl/builtin.py]: Which might mean removing makefile in lots of places.

**Comment by webknjaz** [comment]
This pull request **introduces 3 alerts** when merging 2eb2abec66517c579d314f2c0be087772176f208 into 9f77c31505bf685a74b5536cef8102442bfa5bd0 - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-e8ba8679a9e06512b30d14470ffa8e9e362c9efe)

**new alerts:**

* 2 for Conflicting attributes in base classes
* 1 for Non\-callable called

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 3 alerts** when merging 7b1de00a29980df70e26fa4cb29d0ab513ef3a1f into c92f476b448e025c38911bd9476f7e90ebb144be - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-e861f037a8334836e55b1eb7a873a3fd7828a9c0)

**new alerts:**

* 2 for Conflicting attributes in base classes
* 1 for Non\-callable called

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by webknjaz** [comment]
This pull request **introduces 3 alerts** when merging 8aca957efef4377f150245d7e63f4453d0de16ab into c92f476b448e025c38911bd9476f7e90ebb144be - [view on LGTM.com](https://lgtm.com/projects/g/cherrypy/cheroot/rev/pr-16f93f41b6ceddfdfe3397acdc6197426cda8710)

**new alerts:**

* 2 for Conflicting attributes in base classes
* 1 for Non\-callable called

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Pull request merged**